### ISSUE TITLE
[AdminListBundle] Error saving admin list item that has a File field in a tab

### ISSUE THREAD

**Issue by mtnorthrop** [Kunstmaan/KunstmaanBundlesCMS#1716]
When you have:

1. An admin list item edit form with tabs,
2. A file upload field in one of your tab's form types, but NOT in the form type for the main tab,
3. A File validation constraint on your file upload field...

Then, you get a "File couldn't be found" validation error when you try to upload a file.

It appears that the reason for this is that, since the main form type does not include a file field, the opening `<form>` tag does not automatically include the correct enctype.

A possible solution could be to update the `AdminListBundle/Resources/views/Default/add_or_edit.html.twig` file and check whether a tab pane exists before creating the form with form_start, as was done for the UserManagementBundle in PR #1142.

The `AdminListBundle/Resources/views/Default/add_or_edit.html.twig` file currently reads:
```
{{ form_start(form, {'method': 'POST', 'action': path(app.request.attributes.get('_route'), app.request.attributes.get('_route_params')), 'attr': {'novalidate': 'novalidate'}}) }}
```

It might fix the problem if this was updated to:
```
{% set formParams = {'method': 'POST', 'action': path(app.request.attributes.get('_route'), app.request.attributes.get('_route_params')), 'attr': {'novalidate': 'novalidate'}} %}
{% if tabPane is defined %}
    {{ form_start(tabPane.formView, formParams) }}
{% else %}
    {{ form_start(form, formParams) }}
{% endif %}
```

**Comment by delboy1978uk** [comment]
which version of bundles are you using? I'll try to recreate the issue

**Comment by delboy1978uk** [comment]
Hi again, I have been playing around with this for a while, but I need more info. 

- What is your entity column @ORM set to?
- Have you checked folder permissions?
- Do you use Media type or File type?

Also, I've been told apparently you can set the enctype for the main form in your tab listener, give that a try!

**Comment by mtnorthrop** [comment]
Hello, thanks for looking into this.

In my case, I have a multi-tab form. Each tab has its own Form Type, but all the form types map to the same entity.

The file field in question is on a secondary tab, is a standard Symfony FileType, but is _not_ mapped to Doctrine. (This field is for importing spreadsheets, and the data from the file is parsed before being set to a different property and persisted to the database.)

My workaround for this problem was to include the FileType field in the main tab's FormType, manually set the field as "rendered" in the main tab, and then manually render the field on the tab it belongs on. This lets the AdminBundle know that the form has a FileType and so Symfony renders the correct enctype attribute when `form_start` is called.

I'm not sure where I can force the parent form to render with the correct enctype in the tab listener; from what I've been able to discover, the only way of forcing the enctype is by setting it in the `form_start()` Twig function, but I don't want to override the default AdminListBundle edit template if I can avoid it.

It looks like other users were having the same issue when trying to add tabs to the User form in the UserManagementBundle, and so its edit template was updated to fix the issue, as described in #1142. The second comment by user ro0NL on that ticket addresses the exact same issue I'm having with the AdminListBundle.

My suggestion would be to update all the "tabbable" forms across all the admin bundles to match what's been done in the UserManagementBundle because this would solve the problem and at the same time keep the functionality of all of the bundles' edit forms consistent. However, I don't know whether this could cause BC breaks.

**Comment by delboy1978uk** [comment]
Thanks for the additional info, I'll speak with some of the other devs and see what they think the best solution would be.

**Comment by delboy1978uk** [comment]
We've created a tabbed form macro, and so far implemented it on the view file you mentioned.

As you had a lot going on code wise, could you possibly check out my branch https://github.com/delboy1978uk/KunstmaanBundlesCMS/tree/hotfix/1716 and check it addresses your problem?

If so we'll get it merged in, and we can start looking at where else in the bundles we can use the macro.

Thanks!

**Comment by mtnorthrop** [comment]
Hello,
Thanks for doing this. I'll test this and see if it fixes the problem, but I don't know if I'm going to be able to manage it until the weekend. I should be able to get back to you on Saturday, at the latest.

**Comment by delboy1978uk** [comment]
No rush, let us know when you've gave it a shot, and we'll take it from there.

**Comment by mtnorthrop** [comment]
Hi @delboy1978uk, I've had a look at the patch and it looks like it should work in theory, but there seem to be a couple issues with it. (I'm afraid it didn't work when I tested it.)

One problem is that Twig macros are scoped, so, for example, in
```
{% macro tabbed_form_start(formParams) %}
    {% if tabPane is defined %}
        {{ form_start(tabPane.formView, formParams) }}
    {% else %}
        {{ form_start(form, formParams) }}
    {% endif %}
{% endmacro %}
```
the `tabPane` and `form` variables don't exist. We'd have to pass them into the `tabbed_form_start()` Twig function along with `formParams`. Another option could be to include a Twig template instead of using a macro. When you include Twig templates you have access to the variables in the parent template's scope, unless you specifically indicate otherwise using the `with { foo: 'bar'} only` syntax.

Also, I noticed that you included a `form_end` macro, which looks like a good idea to me, but the current `add_and_edit.html.twig` template just uses a `</form>` tag to end the form. I'm not sure why it was done like this, and it seems to make more sense to use `form_end()`, but maybe there was a reason why just `</form>` was used.

I'd also suggest using the `.html.twig` extension for the `macro_tabbed_form.twig` file, as that's the standard Symfony convention and has been used for all the rest of the Twig macro templates in Kunstmaan.

(I would love to try to fix things like this myself and offer back some actual code, but my project based on Kunstmaan is stuck on 4.0.2 because of the conflict with `symfony-cmf/routing`. My install works thanks to the advice given on https://github.com/Kunstmaan/KunstmaanBundlesCMS/issues/1467#issuecomment-288432710 but I can't do a `composer update` without getting errors.)

Hope this helps!
